File: order-service/src/main/java/com/intexsoft/webshop/orderservice/mapper/DetailMapper.java
package com.intexsoft.webshop.orderservice.mapper;

import com.intexsoft.webshop.messagecommon.event.order.OrderInitializedEvent;
import com.intexsoft.webshop.orderservice.dto.detail.DetailCreateDto;
import com.intexsoft.webshop.orderservice.dto.detail.DetailDto;
import com.intexsoft.webshop.orderservice.model.Detail;
import org.mapstruct.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.Set;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING,
        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
        nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
public interface DetailMapper {
    @Mapping(target = "order", ignore = true)
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    Detail toDetail(DetailCreateDto detailCreateDto);

    @Named("calculateTotalPrice")
    default BigDecimal calculateTotalPrice(DetailCreateDto detailCreateDto) {
        BigDecimal productPrice = detailCreateDto.getProductPrice();
        Short quantity = detailCreateDto.getQuantity();
        return productPrice.multiply(BigDecimal.valueOf(quantity))
                .setScale(2);
    }
    @Mapping(target = "id", source = "id")
    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    DetailDto toDetailDto(Detail detail);

    List<DetailDto> toDetailDtoList(Set<Detail> details);

    List<OrderInitializedEvent.InitOrderProductDetail> toInitOrderProductDetailList(Set<Detail> details);

    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    OrderInitializedEvent.InitOrderProductDetail mapToInitOrderProductDetail(Detail detail);
}

File: order-service-kt/src/main/kotlin/com/intexsoft/webshop/orderservicekt/mapper/DetailMapper.kt
package com.intexsoft.webshop.orderservicekt.mapper

import com.intexsoft.webshop.messagecommon.event.order.OrderInitializedEvent
import com.intexsoft.webshop.orderservicekt.dto.detail.DetailCreateDto
import com.intexsoft.webshop.orderservicekt.dto.detail.DetailDto
import com.intexsoft.webshop.orderservicekt.model.Detail
import org.mapstruct.InjectionStrategy
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    injectionStrategy = InjectionStrategy.CONSTRUCTOR,
)
interface DetailMapper {
    @Mapping(target = "order", ignore = true)
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    fun toDetail(detailCreateDto: DetailCreateDto): Detail

    @Mapping(target = "id", source = "id")
    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    fun toDetailDto(detail: Detail): DetailDto
    fun toDetailDtoList(details: Set<Detail>): List<DetailDto>
    fun toInitOrderProductDetailList(details: Set<Detail>): List<OrderInitializedEvent.InitOrderProductDetail>

    @Mapping(target = "productId", source = "productId")
    @Mapping(target = "productPrice", source = "productPrice")
    @Mapping(target = "quantity", source = "quantity")
    fun mapToInitOrderProductDetail(detail: Detail): OrderInitializedEvent.InitOrderProductDetail
}